slack/rtm/client.py

```diff
@@ -357,18 +366,42 @@ def _next_msg_id(self):
                 raise
 
     async def _read_messages(self):
-        """Process messages received on the WebSocket connection.
-
-        Iteration terminates when the client is stopped or it disconnects.
-        """
+        """Process messages received on the WebSocket connection."""
         while not self._stopped and self._websocket is not None:
-            async for message in self._websocket:
-                if message.type == aiohttp.WSMsgType.TEXT:
-                    payload = message.json()
-                    event = payload.pop("type", "Unknown")
-                    self._dispatch_event(event, data=payload)
-                elif message.type == aiohttp.WSMsgType.ERROR:
-                    break
+            try:
+                # Wait for a message to be recieved, but timeout after a second so that
+                # we can check if the socket has been closed, or if self._stopped is
+                # True
+                message = await self._websocket.receive(timeout=1)
+            except asyncio.TimeoutError:
+                if not self._websocket.closed:
+                    # We didn't receive a message within the timeout interval, but
+                    # aiohttp hasn't closed the socket, so ping responses must still be
+                    # returning
+                    continue
+                self._logger.warning(
+                    "Websocket was closed (%s).", self._websocket.close_code
+                )
+                self._dispatch_event(event="error", data=self._websocket.exception())
+                self._websocket = None
+                self._dispatch_event(event="close")
+                return
+            if message.type == aiohttp.WSMsgType.TEXT:
+                payload = message.json()
+                event = payload.pop("type", "Unknown")
+                self._dispatch_event(event, data=payload)
+            elif message.type == aiohttp.WSMsgType.ERROR:
+                self._logger.error("Received an error on the websocket: %r", message)
+            elif message.type in (
+                aiohttp.WSMsgType.CLOSE,
+                aiohttp.WSMsgType.CLOSING,
+                aiohttp.WSMsgType.CLOSED,
+            ):
+                self._logger.warning("Websocket was closed.")
+                self._websocket = None
+                self._dispatch_event(event="close")
+            else:
+                self._logger.debug("Received unhandled message type: %r", message)
```

Comment:
Same recommendation as above. Let's dispatch the error event if this occurs.
self._dispatch_event(event="error", data=message)

Reply:
I'm not sure receiving an unhandled message type is necessarily an error. It's perfectly fine to ignore those types we don't care about, but I think it's important to log them at debug level in case somebody needs to see exactly what messages are being received over the websocket.

Reply:
That's fair! I initially looked at this as an error but after looking at all of the types specified in WSMsgType I agree that we don't need to handle all of them.